components/dialog/dialog-mixin.js

```diff
@@ -0,0 +1,296 @@
+	}
+
+	_close() {
+		if (!this._state) return;
+		clearDismissible(this._dismissibleId);
+		this._dismissibleId = null;
+		const dialog = this.shadowRoot.querySelector('.d2l-dialog-outer');
+		const transitionEnd = () => {
+			dialog.removeEventListener('transitionend', transitionEnd);
+			if (this._hasNativeDialog) dialog.close();
+			else this._handleClose();
+		};
+		dialog.addEventListener('transitionend', transitionEnd);
+		this._state = 'hiding';
+	}
+
+	_focusOpener() {
+		if (this._opener && this._opener.focus) {
+			this._opener.focus();
+			this._opener = null;
+		}
+	}
+
+	_focusFirst() {
+		const content = this.shadowRoot.querySelector('.d2l-dialog-content');
+		if (content) {
+			const firstFocusable = getNextFocusable(content);
+			if (!firstFocusable.classList.contains('d2l-dialog-trap-end')) {
+				firstFocusable.focus();
+				return;
+			}
+		}
+		this.shadowRoot.querySelector('.d2l-dialog-trap-start').focus();
+	}
+
+	_getHeight() {
+		const availableHeight = window.innerHeight - this._margin.top - this._margin.bottom;
+		let preferredHeight = 0;
+
+		const header = this.shadowRoot.querySelector('.d2l-dialog-header');
+		if (header) preferredHeight += header.scrollHeight;
+
+		const content = this.shadowRoot.querySelector('.d2l-dialog-content > div');
+		if (content) preferredHeight += content.scrollHeight;
+
+		const footer = this.shadowRoot.querySelector('.d2l-dialog-footer');
+		if (footer) preferredHeight += footer.scrollHeight;
+
+		const height = (preferredHeight < availableHeight ? preferredHeight : availableHeight);
+		return height;
+	}
+
+	_getLeft() {
+		if (this._hasNativeDialog || !this._parentDialog) return 0;
+		const parentRect = this._parentDialog.getBoundingClientRect();
+		if (parentRect.width > this._width) return 0;
+		return (parentRect.width - this._width) / 2;
+	}
+
+	_getWidth() {
+		const availableWidth = window.innerWidth - this._margin.left - this._margin.right;
+		const width = (this.width < availableWidth ? this.width : availableWidth);
+		return width;
+	}
+
+	_handleBodyFocus(e) {
+		const dialog = this.shadowRoot.querySelector('.d2l-dialog-outer');
+		const target = e.composedPath()[0];
+		if (isComposedAncestor(dialog, target)) return;
+		this._focusFirst();
+	}
+
+	_handleClose() {
+		/* reset state if native dialog closes unexpectedly. ex. user highlights
+		text and then hits escape key - this is not caught by our key handler */
+		this._removeHandlers();
+		this._focusOpener();
+		this._state = null;
+		this.opened = false;
+		this.dispatchEvent(new CustomEvent(
+			'd2l-dialog-close', { bubbles: true, composed: true }
+		));
+	}
+
+	_handleDialogOpen(e) {
+		this._nestedShowing = true;
+		e.stopPropagation();
+	}
+
+	_handleDialogClose(e) {
+		this._nestedShowing = false;
+		e.stopPropagation();
+	}
+
+	_handleKeyDown(e) {
+		if (!this.opened) return;
+		if (e.keyCode === 27) {
+			// escape (note: prevent native dialog close so we can: animate it; use setDismissible)
```

Comment:
What happens in the case of a nested dialog where this dialog is "open" but isn't the active (top-most) one? Does this handler still get called? Or I guess the top-most one would get called and stop propagation before it gets here.

Reply:
Correct, the innermost (active) dialog would catch it first and `stopPropagation` prevents ancestor dialogs from receiving the key-down event.